### PR TITLE
Generate SQL type definitions for unknown types

### PR DESCRIPTION
**Pull request by weiznich**
This commit adds support for generating SQL type definitions for unknown
types to `diesel print-schema`. The basic idea is to generate the
corresponding marker type to always end up with an existing schema.
Especially this does not generate any code that is required for
serializing/deserializing rust values.

Additionally this commit tweaks the `table!` macro to import types from
the parent scope. This allows us to just reference those newly generated
types easily + simplifies the handling of other custom type imports in
my opinion. The old behaviour of having as part of the `table!`
definition remains supported.

I open this unfinished version of this PR to gather some feedback on the general idea. Before merging this PR we should figure out how to handle this for the mysql backend.

### CODE REVIEW

File: diesel_cli/src/cli.rs [hunk 1]
```diff
@@ -218,6 +218,13 @@ pub fn build_cli() -> App<'static, 'static> {
                 .multiple(true)
                 .number_of_values(1)
                 .help("A list of types to import for every table, separated by commas"),
+        )
+        .arg(
+            Arg::with_name("generate-custom-type-definitions")
```
Comment (pksunkara) [inline diesel_cli/src/cli.rs:223]: Let's name this the same as `generate_missing_sql_type_definitions` and make it a boolean option while defaulting it to `true`. Which means we have to name the option `--no-generate-missing-sql-type-definitions`

**Review by pksunkara — COMMENTED** [review]
I am actually unsure about this change. If my postgres db has an enum `status` and diesel generates a type for it in `schema.rs` file, I would need to edit the schema file to make that generated `Status` struct an actual enum with all possible values. But users are not supposed to edit the schema.rs file.

**Comment by Ten0** [comment]
> If my postgres db has an enum `status` and diesel generates a type for it in `schema.rs` file, I would need to edit the schema file to make that generated `Status` struct an actual enum with all possible values.

That's what I thought too at first glance but then I realized what this generates is the `SqlType` unit struct and not what would be the enum anyway. (The equivalent of what diesel-derive-enum generates as `XxxxxMapping`:
https://github.com/adwhit/diesel-derive-enum/blob/e4286e346fb7063577454cf2aee008f057d20f82/src/lib.rs#L32-L33
https://github.com/adwhit/diesel-derive-enum/blob/e4286e346fb7063577454cf2aee008f057d20f82/src/lib.rs#L206-L210
). You'd still have to declare your enum struct in a separate file, but would only have to implement `FromSql`/`ToSql` & co for it. (Which diesel-derive-enum would have to be slightly updated for.)

Overall it looks like this would remove the large majority of the amount of patch we have to do on our schema, so I think in general this is a very good idea. (By the way now that I look into this I realize [this](https://github.com/diesel-rs/diesel/issues/2484) would have been a much nicer pattern. 😅) I'll take some time later to look at this more in depth and form an opinion about the specific implementation and in particular the naming of the newly generated structs. On that topic, do you have some reference on why it was chosen to not have support for generating rust enums as part of schema inference, apart from https://github.com/diesel-rs/diesel/issues/343#issuecomment-352206453?

**Comment by weiznich** [comment]
> By the way now that I look into this I realize this would have been a much nicer pattern. 😅

It should be quite easy to change the generated schema in such a way that it look like this:
```rust
// this will only be generated if we generate any new sql type
pub mod sql_types {
    pub use diesel::sql_types::*;

    // generated sql types here
}

table! {
    use super::sql_types::*;
    
    foo {
        id -> Integer,
    }
}
```

> particular the naming of the newly generated structs

The naming currently just use the same pattern as previously generated by diesel cli.

> On that topic, do you have some reference on why it was chosen to not have support for generating rust enums as part of schema inference, apart from #343 (comment)?

Basically this + my comment from gitter pasted in the thread below. If the corresponding SQL types are generated third party crates can easily provide an opinionated derive for those use cases.

**Review by Ten0 — COMMENTED** [review]
Looks good :)

## About the naming/scoping of the structs

With the current proposition, `SqlType` structs are named at top level with the sql type name.
This may conflict/not be super intuitive/ergonomic should we later add auto-generation for the rust enums too via the CLI (should it use  [`diesel-derive-enum`](https://github.com/adwhit/diesel-derive-enum/) with `derive`s not directly implemented in Diesel).

### Why it may be important

> Basically [this](https://github.com/diesel-rs/diesel/issues/343#issuecomment-352206453) + [my comment from gitter pasted in the thread below](https://github.com/diesel-rs/diesel/issues/343#issuecomment-625712565). If the corresponding SQL types are generated third party crates can easily provide an opinionated derive for those use cases.

The advantage of having this as part of Diesel would be that in order to generate the rust enums by querying the database schema (to prevent mismatches between enum declarations in Rust and in SQL and reduce boilerplate, same reason why the Diesel CLI exists...), if this is an external crate, we would need an additional CLI , while this would otherwise easily be integrated as part of the existing Diesel CLI.

Considering the existing crate that serves this purpose: https://github.com/adwhit/diesel-derive-enum/blob/e4286e346fb7063577454cf2aee008f057d20f82/src/lib.rs#L81-L103 it looks like:
1. Such a feature would actually not have to be harshly opinionated: it could be a setting of the CLI.
2. Creating and maintaining an additional CLI seems to not have been worth the hassle enough that someone wanted to do it.

So anyway, not saying we should do it right now, but although this may not have been something that was appropriate to support [four years ago](https://github.com/diesel-rs/diesel/issues/343#issuecomment-352206453), it might eventually become something we want to do and I think it's best if we leave the possibility for the structs namings to still be sound should we later decide to add support for it.

### Solving

The design described at https://github.com/diesel-rs/diesel/pull/2787#discussion_r645894655 solves this issue.

File: diesel/src/macros/mod.rs [hunk 2]
```diff
@@ -616,10 +616,10 @@ macro_rules! __diesel_table_impl {
         },)+],
     ) => {
         $($meta)*
+        #[allow(unused_imports)]
         pub mod $table_name {
-            #![allow(dead_code)]
-            $($imports)*
             pub use self::columns::*;
+            $($imports)*
```
[the remark this answers is not in the document]
Comment (Ten0) [reply diesel/src/macros/mod.rs:622]: nvm, got it: https://github.com/diesel-rs/diesel/pull/2787#discussion_r645894655

File: diesel_cli/tests/print_schema/print_schema_default_is_to_generate_custom_types/postgres/expected.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+/// The `MyType` SQL type
+///
+/// (Automatically generated by Diesel.)
+#[derive(diesel::SqlType)]
+#[postgres(type_name = "my_type")]
+pub struct MyType;
+
+diesel::table! {
+    use diesel::sql_types::*;
+    use super::MyType;
```
[the remark this answers is not in the document]
Comment (weiznich) [reply diesel_cli/tests/print_schema/print_schema_default_is_to_generate_custom_types/postgres/expected.rs:12]: Putting the types into a separate module is a great idea. I will change that as soon as I find some time to work on this again. That written I fear we cannot just automatically import a parent `sql_types` module here behind the scenes as this module may not exists for hand written `table!` macro calls and I would prefer not to require that module in that cases.

**Comment by weiznich** [comment]
> Why it may be important
>
>>    Basically this + my comment from gitter pasted in the thread below. If the corresponding SQL types are generated third party crates can easily provide an opinionated derive for those use cases.
>
> The advantage of having this as part of Diesel would be that in order to generate the rust enums by querying the database schema (to prevent mismatches between enum declarations in Rust and in SQL and reduce boilerplate, same reason why the Diesel CLI exists...), if this is an external crate, we would need an additional CLI , while this would otherwise easily be integrated as part of the existing Diesel CLI.

It may be reasonable to add this to diesel itself in some future version, but I would prefer to the already low amount of time I can spend on working on diesel. So this is some feature I would put onto the "Implement this someday if there are more maintainers" list. That written, we should definitively design this feature in such a way that it enables this future extension.

File: diesel_cli/src/print_schema.rs [hunk 4]
```diff
@@ -1,14 +1,16 @@
...
 use serde::de::{self, MapAccess, Visitor};
 use serde::{Deserialize, Deserializer, Serialize};
 use serde_regex::Serde as RegexWrapper;
+use std::collections::HashSet;
 use std::error::Error;
 use std::fmt::{self, Display, Formatter, Write};
 use std::io::Write as IoWrite;
 
-const SCHEMA_HEADER: &str = "// @generated automatically by Diesel CLI.\n";
+const SCHEMA_HEADER: &str = "// @generated automatically by Diesel CLI.";
```
[the remark this answers is not in the document]
Comment (weiznich) [reply diesel_cli/src/print_schema.rs:13]:
```suggestion
const SCHEMA_HEADER: &str = "// @generated automatically by Diesel CLI.\n";
```

File: diesel_cli/src/print_schema.rs [hunk 5]
```diff
@@ -78,20 +168,54 @@ pub fn output_schema(
     let table_data = table_names
         .into_iter()
         .map(|t| load_table_data(database_url, t, &config.column_sorting))
-        .collect::<Result<_, Box<dyn Error + Send + Sync + 'static>>>()?;
+        .collect::<Result<Vec<_>, Box<dyn Error + Send + Sync + 'static>>>()?;
+
+    let mut out = String::new();
+    writeln!(out, "{}", SCHEMA_HEADER)?;
+
+    writeln!(out)?;
```
Comment (weiznich) [inline diesel_cli/src/print_schema.rs:176]:
```suggestion
```

File: diesel_cli/src/print_schema.rs [hunk 6]
```diff
@@ -187,10 +354,26 @@ impl<'a> Display for TableDefinition<'a> {
...
                     writeln!(out, "use {};", import)?;
+                    has_written_import = true;
+                }
+            }
+
+            for col in &self.table.column_data {
+                dbg!(&col.rust_name);
+                if dbg!(self.custom_type_defs.contains(&col.ty.rust_name)) {
+                    if !has_written_import {
+                        writeln!(out, "use diesel::sql_types::*;")?;
```
[the remark this answers is not in the document]
Comment (weiznich) [reply diesel_cli/src/print_schema.rs:409]: Yes that's correct

**Pull request merged**